### PR TITLE
[FIX] website_crm_partner_assign: reseller localization

### PR DESCRIPTION
**Pull request by nd-dew**
### Issue Description:
The partner() function previously inferred the country using GeoIP
when no specific country (nor "all_countries") was specified.
However, if there were no partners from the inferred country,
the render function would lead to a broken screen.

### Resolution:
Implemented a fallback mechanism to treat the situation
as if no country was specified. This ensures that all partners are displayed
in such cases, preventing the occurrence of a broken screen.

### [Reproducing bug]
- odoo 16.4 with: website, contacts, crm_enterprise_partner_assign
- Simulate GeoIP on the backend, so it points to some COUNTRY_A:
    - this can be achived by adding `country= country_obj.search([('name', 'ilike', 'mex')])` [below this line](https://github.com/odoo/odoo/blob/57d43aca2b07d3469a937890b7d5e5c4f29fe250/addons/website_crm_partner_assign/controllers/main.py#L237)
    - note: it doesn't matter which country you set it to, AS LONG AS COUNTRY_A != COUNTRY_B
- in CRM/Contacts:
   - create partner (set Partner Level and country to COUNTRY_B of the record )
   - Publish partner by "Go to website" -> Published
- go to url: ODOO_WEBSITE/partners
- BUG: No Partners are displayed

opw-3598110

### CODE REVIEW

**Comment by nd-dew** [comment]
That how broken screen look like.
image

**Comment by nd-dew** [comment]
**Divagations on the test:**
 I check situation where end-user has location "mexico",  and no partners are from mexico, but I don't enforce that (there are no partners from mexico), should I ? Namely, should i find and delete all partners from Mexico, in case some future changes will change the TestSetup class?
 
 Should I test various locations (not only "MX") ? ( I don't see anything mexico specific in the logic of the tested functionality. So i don't think it is necessary.)

File: addons/test_website/tests/test_fuzzy.py [hunk 1]
```diff
@@ -96,3 +98,34 @@ def test_02_pages_search(self):
...
+            Patching is allowing to modify normally read_only value.
+        """
+        # Patch GeoIp so it acts, as if Odoo client is located in Mexico
+        GeoIpMock.return_value.country_code = 'MX'
+
+        # Create a partner outside of Mexico
+        non_mexican_partner = self.env['res.partner'].create({
+            'name': 'Non_Mexican_Partner',
+            'is_company': True,
+            'grade_id': self.env['res.partner.grade'].search([], limit=1).id,
```
Comment (nle-odoo) [inline addons/test_website/tests/test_fuzzy.py:119]: #this model is added by website_crm_partner_assign so you can't use it in test_website (you can only use it in website_crm_partner_assign module or module that depends on website_crm_partner_assign)

same thing for `WebsiteCrmPartnerAssign` it's only usable those modules too

File: addons/website_crm_partner_assign/controllers/main.py [hunk 2]
```diff
@@ -250,13 +265,8 @@ def partners(self, country=None, grade=None, page=0, **post):
...
         })
 
-        # group by country
-        country_domain = list(base_partner_domain)
-        if grade:
-            country_domain += [('grade_id', '=', grade.id)]
-        countries = partner_obj.sudo().read_group(
-            country_domain, ["id", "country_id"],
-            groupby="country_id", orderby="country_id")
+
```
Comment (nle-odoo) [inline addons/website_crm_partner_assign/controllers/main.py:268]: too empty lines spaces (and also one line too little before `@patch('odoo.http.GeoIP')`

**Review by nle-odoo — APPROVED** [review]
looks good to me

side note: besides moving `group by country` the only new lines are:

```py
# Fallback: Show all partners when country has no associates.
country_ids = [c['country_id'][0] for c in countries]
if country and country.id not in country_ids:
    country = None
```

File: addons/test_website/tests/test_fuzzy.py (hunk 1 above)
Comment (nd-dew) [reply addons/test_website/tests/test_fuzzy.py]: good point! how about moving the test to `website_crm_partner_assign` module ?

**Comment by nd-dew** [comment]
Added assert making sure that render function was called.

**Review by Julien00859 — COMMENTED** [review]
that GeoIP mock in the unittest is neat but I know nothing related to the flow you're fixinx

**Review by reth-odoo — APPROVED** [review]
Yeah, seems fine to me too, thanks for the fix :)

Seems like an odd behavior to have in the first place, I would have thought we'd set the filter in the front-end if relevant and otherwise not do any filtering by default... But it is what it is, looks llike it's historic code.

**Comment by nd-dew** [comment]
@robodoo r+

**Comment by nd-dew** [comment]
can I get r+

**Comment by nle-odoo** [comment]
robodoo r+

**Pull request closed without merging**